### PR TITLE
New data sources

### PR DESCRIPTION
**Pull request by SwampDragons**
Merge after #11247 

This PR creates two new data sources  -- hcp-packer-iteration and hcp-packer image. They are meant to be used together to query HCP. 

Usage example: 

```
source "null" "example" {
  communicator = "none"
}

data "hcp-packer-iteration" "hardened-source" {
  bucket_name = "hardened-ubuntu-16-04"
  channel = "megan-test"
}

data "hcp-packer-image" "aws" {
  bucket_name = "hardened-ubuntu-16-04"
  iteration_id = "${data.hcp-packer-iteration.hardened-source.id}"
  cloud_provider = "aws"
  region = "us-east-1"
}

locals {
  foo              = "${data.hcp-packer-iteration.hardened-source.id}"
  bar              = "${data.hcp-packer-image.aws.image_id}"
}

build {
  name = "mybuild"
  sources = [
    "source.null.example"
  ]
  provisioner "shell-local" {
    inline = [
      "echo data is ${local.foo}",
      "echo data is ${local.bar}"
    ]
  }
}

```

Note that since this example requires data sources to be able to use other data sources, it won't work until my next Pr is merged (link incoming once I make the PR)

### CODE REVIEW

File: datasource/hcp-packer-image/data.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	var errs *packersdk.MultiError
+
+	if d.config.Bucket == "" {
+		errs = packersdk.MultiErrorAppend(errs, fmt.Errorf("The `bucket_name` must be specified"))
+	}
+	if d.config.IterationID == "" {
+		errs = packersdk.MultiErrorAppend(errs, fmt.Errorf("The `iteration_id`"+
+			" must be specified. If you do not know your iteration_id, you "+
+			"can retrieve it using the bucket name and desired channel using"+
+			" the hcp-packer-iteration provider"))
```
Comment (sylviamoss) [inline datasource/hcp-packer-image/data.go:59]: nice!

File: datasource/packer-image-iteration/data.go [hunk 2]
```diff
@@ -122,7 +122,7 @@ type DatasourceOutput struct {
 	CreatedAt string `mapstructure:"created_at"`
 	// A list of builds that are stored in the iteration. These builds can be
 	// parsed using HCL to find individual image ids for specific providers.
-	Builds []ParBuild `mapstructure:"builds"`
+	Builds map[string]ParBuild `mapstructure:"builds"`
```
Comment (sylviamoss) [inline datasource/packer-image-iteration/data.go:125]: Are these changes needed yet?

Comment (SwampDragons) [reply datasource/packer-image-iteration/data.go:125]: ah, must have accidentally re-added them in a squash. I'll clean up now.

File: command/plugin.go [hunk 3]
```diff
@@ -60,6 +62,8 @@ var PostProcessors = map[string]packersdk.PostProcessor{
 }
 
 var Datasources = map[string]packersdk.Datasource{
+	"hcp-packer-image":       new(hcppackerimagedatasource.Datasource),
+	"hcp-packer-iteration":   new(hcppackeriterationdatasource.Datasource),
```
Comment (azr) [inline command/plugin.go:66]: Could have been nice to make two PRs here !

**Review by azr — APPROVED** [review]
Nice !
